### PR TITLE
Ipfs shields

### PR DESCRIPTION
**Pull request by bbondy**
Resolves https://github.com/brave/brave-browser/issues/13741
Resolves https://github.com/brave/brave-browser/issues/13742

Reviewing per commit is easiest.

Brave extension sees IPFS URIs as ipfs://.  Instead if we let it see URIs as what they resolve to, it makes everything work correctly.  Furthermore, after the brave extension issue is fixed, adblock was still not working because the tab URL was always blank.  This sets it to the initiator URI which is actually available. 

## Submitter Checklist:

- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed).
- [ ] Requested a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by yrliou — APPROVED** [review]
Overall looks good, just a nit.

File: browser/extensions/api/ipfs_api.cc [hunk 1]
```diff
@@ -146,5 +155,30 @@ ExtensionFunction::ResponseAction IpfsGetExecutableAvailableFunction::Run() {
...
+      ipfs::ResolveIPFSURI::Params::Create(*args_));
+  EXTENSION_FUNCTION_VALIDATE(params.get());
+  GURL uri(params->uri);
+  Profile* profile = Profile::FromBrowserContext(browser_context());
+  auto* prefs = profile->GetPrefs();
+  bool local = static_cast<IPFSResolveMethodTypes>(prefs->GetInteger(
+                   kIPFSResolveMethod)) == IPFSResolveMethodTypes::IPFS_LOCAL;
+  GURL base_gateway_url(
+      local ? ::ipfs::GetDefaultIPFSLocalGateway(chrome::GetChannel())
+            : ::ipfs::GetDefaultIPFSGateway(browser_context()));
```
[the remark this answers is not in the document]
Comment (bbondy) [reply browser/extensions/api/ipfs_api.cc:173]: I did do that at first and then realized I needed to call `chrome::GetChannel()` which requires `chrome/` so I decided not to do that refactor. But I guess we can just pass in the value.

File: components/ipfs/translate_ipfs_uri.cc [hunk 2]
```diff
@@ -42,13 +46,24 @@ bool TranslateIPFSURI(const GURL& url, GURL* new_url, const GURL& gateway_url) {
...
+          std::string host = base::StringPrintf("%s.%s.%s", cid.c_str(),
+                                                ipfs_scheme ? "ipfs" : "ipns",
+                                                gateway_url.host().c_str());
+          replacements.SetHostStr(host);
+          replacements.SetPathStr(path);
+          *new_url = gateway_url.ReplaceComponents(replacements);
+        } else {
+          std::string gateway_empty_path = gateway_url.spec();
+          if (gateway_empty_path.size() > 0 &&
+              gateway_empty_path[gateway_empty_path.size() - 1] == '/') {
```
Comment (iefremov) [inline components/ipfs/translate_ipfs_uri.cc:61]: this is `gateway_empty_path.back()`

File: components/ipfs/translate_ipfs_uri.h [hunk 3]
```diff
@@ -10,7 +10,10 @@
 
 namespace ipfs {
 
-bool TranslateIPFSURI(const GURL& url, GURL* new_url, const GURL& gateway_url);
+bool TranslateIPFSURI(const GURL& url,
+                      GURL* new_url,
+                      const GURL& gateway_url,
+                      bool use_subdomain);
```
Comment (iefremov) [inline components/ipfs/translate_ipfs_uri.h:16]: IMO using another function like `TranslateIPFSURIWithSubdomain` would be more readable (and there is only one spot where it is needed), but not a big deal

**Review by iefremov — APPROVED** [review]

File: components/ipfs/translate_ipfs_uri.cc (hunk 2 above)
Comment (bbondy) [reply components/ipfs/translate_ipfs_uri.cc:61]: that code is removed in the rebase on master I'll be pushing soon, so no worries.

**Review by yrliou — APPROVED** [review]
LGTM

File: components/ipfs/translate_ipfs_uri.h (hunk 3 above)
Comment (bbondy) [reply components/ipfs/translate_ipfs_uri.h]: Will consider on a future refactor, thanks!

**Pull request merged**